clustering/ejb/WFLY-7628_Distributed_EJB_Timers.adoc

```diff
@@ -0,0 +1,168 @@
+= Distributed EJB timers
+:author:            Paul Ferraro
+:email:             paul.m.ferraro@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Currently, WildFly supports 2 mechanisms for persistent EJB timers: a file-based implementation and a JDBC-based implementation.
+Of these, only the JDBC-based implementation is suitable for use in a cluster.
+However, this implementation does not scale well in large to medium sized clusters and relies on a single point of failure due to the blocking nature of the mechanism employed to determine which cluster member should schedule a given timer expiration.
+WildFly needs an persistent timer service implementation suitable for use in a cluster that is both highly-available and reasonably scalable.
+This feature will add a new timer service implementation that uses the consistent hash of an Infinispan cache to determine which timers a given cluster member should manage.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-7628[WFLY-7628]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/WFLY-14953[WFLY-14953]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:msvehla@redhat.com[Martin Svehla]
+* mailto:fburzigo@redhat.com[Fabio Burzigotti]
+
+=== Testing By
+
+* [ ] Engineering
+
+* [x] QE
+
+=== Affected Projects or Components
+
+Clustering, EJB
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+* Create a `TimerService` implementation that delegates to separate persistent vs transient `TimerService` implementations.
+* Implement Infinispan-based `TimerService` for persistent timers that manages timer execution based on the consistent hash of the timer identifier.
+** Make existing ``default-data-store`` attribute of the timer service resource optional.  Use distributable implementation when this attribute is undefined.
+* Add default and named persistent timer management capabilities to distributable-ejb subsystem.
+** Add named Infinispan timer management resource that provides the above capability.
+* Add an attribute to the timer service resource within the ejb3 subsystem to determine the server-wide default provider for transient timer management.
+* Add ability to enable distributed timer management (in lieu of a timer store) within ejb3 subsystem for management of persistent timers.
+* Existing timer service resource operations must continue to work when executed against the distributable `TimerService`.
+
+=== Nice-to-Have Requirements
+
```

Comment:
All the nice-to-have requirements deal with transient timers, although no tests, between those mentioned in this analysis document, are defined to check such requirement configuration and functionalities, or am I missing something?

Reply:
The same integration test will cover both cases.   I have not yet committed these changes.

Reply:
I see, thanks. I'll keep an eye on the feature implementation PR and we can update this very thread accordingly until resolved.

Reply:
The feature PR includes these tests now.

Reply:
Thanks.